doc/design/diagnosis.md

```diff
@@ -0,0 +1,82 @@
+# SQLFlow Diagnosis
+
+For a typical SQLFlow method calling link:
+
+``` golang
+// pkg/sqlflowserver
+func Do(program string, sess *Session) string {
+  return workflow.Run(
+    couler.GenCode(
+      verifier.Verify(
+        resolver.ResolveProgram(parser.Parse(program)),
+        sess.DB)))
+}
+```
+
+Any of the above function calls can raise an error, it may be a syntax error on parsing, or verify fault on verifying prediction and training data schema, or the submitter program running error, etc.
+
+To improve the error message maintainability and readability, this document issues an SQLFlow Error package named `sferr` to maintain various error messages in a rigorous approach.
+
+## Error Package
+
+We would like each Go package under `pkg` folder do one minimal thing, for each package, we represent different `ErrType`, and using `ErrCode` to
+distinguish different errors for the same `ErrType`, the error code can be the same among all `ErrType`s.
+
+``` golang
+// ErrType represents a specific Error Type
+type ErrType int
+
+// ErrCode represents a specific Error Type in a ErrType
+type ErrCode int
+
+type Error struct {
```

Comment:
How about making this type compliant with the standard `Error` interface? Then we can use this type everywhere an `Error` is needed.

Reply:
Done.